waiter/src/waiter/auth/oidc.clj

```diff
@@ -0,0 +1,211 @@
+   (preserved in the state) before the auth flow was triggered.
+   Unsuccessful authentication returns either a 400 Bad request, the downstream auth server
+   response, or a 401 unauthorized with appropriate details."
+  [{:keys [jwt-auth-server jwt-validator password] :as oidc-authenticator} request]
+  (if (nil? oidc-authenticator)
+    (utils/clj->json-response {:message "OIDC authentication disabled"} :status http-501-not-implemented)
+    (let [{:keys [code-verifier code state-map]} (validate-oidc-callback-request password request)]
+      (async/go
+        (let [access-token-ch (jwt/request-access-token jwt-auth-server request oidc-callback-uri code code-verifier)
+              body-or-throwable (async/<! access-token-ch)]
+          (try
+            (if (instance? Throwable body-or-throwable)
+              (throw body-or-throwable)
+              (let [access-token body-or-throwable
+                    _ (log/info "successfully retrieved access token" (utils/truncate access-token 30))
+                    key-id->jwk (jwt/get-key-id->jwk jwt-auth-server)
+                    result-map-or-throwable (jwt/extract-claims jwt-validator key-id->jwk request access-token)
+                    _ (when (instance? Throwable result-map-or-throwable)
+                        (throw result-map-or-throwable))
+                    {:keys [expiry-time subject]} result-map-or-throwable
+                    _ (log/info "authenticated subject is" subject)
+                    auth-params-map (auth/build-auth-params-map :oidc subject {:jwt-access-token access-token})
+                    auth-cookie-age-in-seconds (- expiry-time (jwt/current-time-secs))]
+                (auth/handle-request-auth
+                  (constantly
+                    (let [{:keys [redirect-uri]} state-map]
+                      (-> {:headers {"cache-control" "no-store"
+                                     "content-security-policy" content-security-policy-value
+                                     "location" redirect-uri}
+                           :status http-302-moved-temporarily}
+                        (cookie-support/add-encoded-cookie password oidc-challenge-cookie "" 0)
+                        (utils/attach-waiter-source))))
+                  request auth-params-map password auth-cookie-age-in-seconds)))
+            (catch Throwable throwable
+              (log/info throwable "error in retrieving access token")
+              (utils/exception->response
+                (ex-info (.getMessage throwable)
+                         (-> (ex-data throwable)
+                           (utils/assoc-if-absent :log-level :info)
+                           (utils/assoc-if-absent :status http-401-unauthorized))
+                         throwable)
+                request))))))))
+
+(defn trigger-authorize-redirect
+  "Triggers a 302 temporary redirect response to the authorize endpoint."
+  [jwt-auth-server password {:keys [query-string uri] :as request} response]
```

Comment:
In what critical ways is this a function of the existing response?

Reply:
It needs to preserve any of the response headers (e.g. `www-negotiate` headers) and other keys mapped (e.g. `source`) in the response.